werkzeug/routing.py

```diff
@@ -248,11 +249,33 @@ class BuildError(RoutingException, LookupError):
     values provided.
     """
 
-    def __init__(self, endpoint, values, method):
+    def __init__(self, endpoint, values, method, suggested_rule):
         LookupError.__init__(self, endpoint, values, method)
         self.endpoint = endpoint
         self.values = values
         self.method = method
+        self.suggested_rule = suggested_rule
+
+    def __str__(self):
+        message = []
+        message.append("Could not build url for endpoint %r" % self.endpoint)
+        if self.method:
+            message.append(" (%r)" % self.method)
+        if self.values:
+            message.append(" with values %r" % self.values.keys())
```

Comment:
This is broken with Python 3, it shows the repr of an iterator. I suggest wrapping with ``list(...)``, and if you want to avoid recreating the list, use ``werkzeug._compat.iterkeys`` instead of ``dict.keys``.